lib/tbot/service/service.go

```diff
@@ -0,0 +1,186 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package service
+
+import (
+	"context"
+	"errors"
+	"slices"
+	"sync/atomic"
+
+	"github.com/gravitational/teleport/lib/tbot/service/status"
+)
+
+// Handler implements the logic of a long-running service.
+type Handler interface {
+	// Run performs the task (potentially on an interval) until the given
+	// context is canceled, or an irrecoverable error is encountered.
+	//
+	// It should call runtime.SetStatus to report its health to the supervisor
+	// and other services. If Run returns an error, the service's status will be
+	// set to Failed and the supervisor will restart the handler after some
+	// backoff and jitter.
+	//
+	// If the handler encounters an error that cannot possibly be resolved, it
+	// should return the error wrapped with IrrecoverableError and the supervisor
+	// will shut down all of the other services.
+	Run(ctx context.Context, runtime *Runtime) error
+}
+
+// OneShotHandler can be implemented by a service to support running in one-shot
+// mode rather than as a long-running daemon.
+type OneShotHandler interface {
+	// OneShot performs the task once and then exits. If it returns nil, the
+	// service's status will be set to Ready otherwise it will be set to Failed.
+	OneShot(ctx context.Context) error
+}
```

Comment:
Wouldn't it be possible & simpler to have `Handler.IsOneShot() bool` instead?

Reply:
Just to check I understand: do you mean handlers having a single `Run` method and an `IsOneShot` method to determine whether it should be run in long-running or one-shot mode?

Currently, services are `long-running || (long-running && one-shot)` 😅 as in they either _only_ implement long-running mode or _both_ long-running and one-shot.

Arguably if we moved the "run this on an interval" logic into the supervisor, we probably could just give handlers the one method and a way to mark themselves as being long-running only (e.g. the HTTP diagnostics service).

WDYT @strideynet? 🤔

Reply:
Yeah - I think for the most part, we do have significantly more services that are "oneshot on an interval" as opposed to truly long-running so there may some value in that. At the moment we don't have a lot of uniformity between these and this could be useful in bringing some.

I think we'd just want to be careful that we offer enough "control" over behaviour because something like the Bot identity service is a little different to the rest - we'd also want to make sure that we don't overly optimize for the "oenshot on an interval" services because I expect we'll have more real long-running services coming in the next few months.

Reply:
I'm quite drawn to the idea of keeping `Service[T]` as lowest-level of abstraction, and providing a layer on top for the common "oneshot on an interval" case.

For example:

```Go
fooService := service.New(
	"foo",
	service.IntervalHandler(&fooHandler{}, 1*time.Minute)
)

type fooHandler struct{}

func (fooHandler) Run(ctx context.Context) error { return nil }
```

Where `IntervalHandler` would return a wrapper that implements `service.Handler`, calls the given handler on an interval, and automatically sets the service's status for you.

Reply:
> Currently, services are `long-running || (long-running && one-shot)` 😅 as in they either _only_ implement long-running mode or _both_ long-running and one-shot.

This notion is not clear to me at all from the interface. Maybe extending the godoc could help.